langlib/lang.value/src/main/java/org/ballerinalang/langlib/value/FromJsonWithType.java

```diff
@@ -0,0 +1,268 @@
+            throw createConversionError(value, targetType);
+        } else if (convertibleTypes.size() > 1) {
+            throw createConversionError(value, targetType, AMBIGUOUS_TARGET);
+        }
+
+        BType matchingType = convertibleTypes.get(0);
+
+        Object newValue;
+        switch (sourceType.getTag()) {
+            case TypeTags.MAP_TAG:
+            case TypeTags.RECORD_TYPE_TAG:
+                newValue = convertMap((MapValue<?, ?>) value, matchingType, unresolvedValues, t, strand);
+                break;
+            case TypeTags.ARRAY_TAG:
+            case TypeTags.TUPLE_TAG:
+                newValue = convertArray((ArrayValue) value, matchingType, unresolvedValues, t, strand);
+                break;
+            case TypeTags.ERROR_TAG:
+                newValue = ((RefValue) value).copy(new HashMap<>());
+                break;
+            default:
+                if (((matchingType.getTag() >= TypeTags.XML_ELEMENT_TAG &&
+                        matchingType.getTag() <= TypeTags.XML_TEXT_TAG) || matchingType.getTag() == TypeTags.XML_TAG)) {
+                    try {
+                        newValue = XMLFactory.parse(((StringValue) value).getValue());
+                        break;
+                    } catch (Throwable e) {
+                        throw createConversionError(value, targetType, e.getMessage());
+                    }
+                }
+
+                // handle primitive values
+                if (sourceType.getTag() <= TypeTags.BOOLEAN_TAG) {
+                    if (TypeChecker.checkIsType(value, matchingType)) {
+                        newValue = value;
+                    } else {
+                        // Has to be a numeric conversion.
+                        newValue = TypeConverter.convertValues(matchingType, value);
+                    }
+                    break;
+                }
+                // should never reach here
+                throw BallerinaErrors.createConversionError(value, targetType);
+        }
+
+        unresolvedValues.remove(typeValuePair);
+        return newValue;
+    }
+
+    private static Object convertMap(MapValue<?, ?> map, BType targetType, List<TypeValuePair> unresolvedValues,
+                                     TypedescValue t, Strand strand) {
+        switch (targetType.getTag()) {
+            case TypeTags.MAP_TAG:
+                MapValueImpl<BString, Object> newMap = new MapValueImpl<>(targetType);
+                for (Map.Entry entry : map.entrySet()) {
+                    BType constraintType = ((BMapType) targetType).getConstrainedType();
+                    putToMap(newMap, entry, constraintType, unresolvedValues, t, strand);
+                }
+                return newMap;
+            case TypeTags.RECORD_TYPE_TAG:
+                BRecordType recordType = (BRecordType) targetType;
+                MapValueImpl<BString, Object> newRecord;
+                if (t != null && t.getDescribingType() == targetType) {
```

Comment:
Do we need this null check here? `t` can never be `null` right?

Reply:
Ack